### PR TITLE
Analysis #6

### PR DESCRIPTION
**Pull request by maggiesu0725**
Set up NCHousing_needs_units_targets and NCHousing_needs_units_targets-Alt

### CODE REVIEW

**Comment by maggiesu0725** [comment]
@lhendey please review Hud_inc_NCState.sas the macro and the main program that creates the dataset NCHousing_needs_units_targets.sas  There is one metro2013 value (34820) that's not in the income limits spreadsheet that I need to find out why. If you could pay extra attention to the different weights that would be great, I got confused and not confused and confused again during the programming.

**Comment by lhendey** [comment]
@maggiesu0725 i'll look at it in the morning.

**Comment by lhendey** [comment]
@maggiesu0725 i got stuck on the household income program. i don't think the crosswalk merge is working right for all jurisdictions. can we discuss?

File: Prog/Tabulate householder income for projection.sas [hunk 1]
```diff
@@ -158,14 +158,34 @@ run;
...
+%tabulateinc(2017);*/
 
-/*compile 13-17 data for tabulation */
-data fiveyeartotal;
-set Householddetail_2013_inc  Householddetail_2014_inc Householddetail_2015_inc Householddetail_2016_inc Householddetail_2017_inc;
+data fiveyeartotal1;
+set Householddetail_2013  Householddetail_2014 Householddetail_2015 Householddetail_2016 Householddetail_2017;
 totalpop=0.2;
+merge=1;
 totpop_wt= totalpop*AFACT2;
```
Comment (lhendey) [inline Prog/Tabulate householder income for projection.sas]: remove afact2 from this program and the householder population program.

**Comment by lhendey** [comment]
@maggiesu0725 okay to look at the other programs now?

**Comment by maggiesu0725** [comment]
yes. @lhendey

File: Prog/NCHousing_needs_units_targets.sas [hunk 2]
```diff
@@ -0,0 +1,927 @@
...
+    4 = '81-120%'
+    5 = '120-200%'
+    6 = 'More than 200%'
+	;
+
+  value tenure
+    1 = 'Renter units'
+    2 = 'Owner units'
+	;
+/*
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: just remove this.

File: Prog/NCHousing_needs_units_targets.sas [hunk 3]
```diff
@@ -0,0 +1,927 @@
...
+	length puma_new $5;
+	puma_new = put(input(cats(puma12),8.),z5.);
+	upuma= "37"||puma_new;
+run;
+
+proc sort data= crosswalk;
+	by upuma;
+run;
+
+data NC17_limits;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: you can actual do this with one step. 

proc sort data=NCHsg.NC17limits out=NC17_limits;
by MID;
run;

File: Prog/NCHousing_needs_units_targets.sas [hunk 4]
```diff
@@ -0,0 +1,927 @@
...
+	if a;
+	by upuma ;
+	%assign_NCcounty2;
+	%assign_NCcounty3;
+	county2_char = county2;
+
+	run;
+
+
+	data NCarea_&year._1 ;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: same comment as above- this data step and proc sort can be combined in to one proc sort with out=.

File: Macros/Hud_inc_NCState.sas [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+ 5  =  120-200% AMI (high)
+ 6  =  >=200% (extremely high)
+ -99  =  N/A (income not reported)
+
+ Modifications: Yipeng Su from Hud_inc_RegHSG for NCHsg project.
+**************************************************************************/
+
+/** Macro Hud_inc_RegHsg - Start Definition **/
+
+%macro Hud_inc_NCState(hhinc=, hhsize=  );
```
Comment (lhendey) [inline Macros/Hud_inc_NCState.sas]: this looks good, but please add category 6 (gt 200% of AMI)

File: Prog/NCHousing_needs_units_targets.sas [hunk 6]
```diff
@@ -0,0 +1,927 @@
...
+
+    
+    ****** Rental units ******;
+    
+   if ownershpd in (21, 22) then do;
+        
+    Tenure = 1;
+
+	 *create maximum desired or affordable rent based on HUD_Inc categories*; 
+    /* need to discuss for NC, use hudinc for now*/
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: you need to compute the averages for NC for each hud_income group to determine % of income for hud_inc = 4, 5, and 6.

File: Prog/NCHousing_needs_units_targets.sas [hunk 7]
```diff
@@ -0,0 +1,927 @@
...
+
+		if max_rent ~= . then do; 
+			if max_rent > rentgrs_a*1.1 then couldpaymore=1; 
+			else if max_rent <= rentgrs_a*1.1 then couldpaymore=0; 
+		end; 
+
+	
+    	*rent cost categories that make more sense for rents - no longer used in targets;
+		/*need to discuss for NC*/
+			rentlevel=.;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: these don't look right to me. i think you want to start the categories much lower  - 30% of income for an ELI renter (statewide) is $443. the categories don't have to precisely align with AMI but the bottom 3 categories should be close to 0-30% of AMI, 30-50% and 50-80%.

File: Prog/NCHousing_needs_units_targets.sas [hunk 8]
```diff
@@ -0,0 +1,927 @@
...
+			if 750 <=max_rent<1200 then mrentlevel=2;
+			if 1200 <=max_rent<1500 then mrentlevel=3;
+			if 1500 <=max_rent<2000 then mrentlevel=4;
+			if 2000 <=max_rent<2500 then mrentlevel=5;
+			if max_rent >= 2500 then mrentlevel=6;
+
+		 *rent cost categories now used in targets that provide a set of categories useable for renters and owners combined; 
+			/*need to discuss for NC*/
+			allcostlevel=.;
+			if rentgrs_a<800 then allcostlevel=1;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: see comment above.

File: Prog/NCHousing_needs_units_targets.sas [hunk 9]
```diff
@@ -0,0 +1,927 @@
...
+  	else if ownershpd in ( 12,13 ) then do;
+
+	    ****** Owner units ******;
+	    
+	    Tenure = 2;
+
+		*create maximum desired or affordable owner costs based on HUD_Inc categories*; 
+
+		/*need to discuss for NC*/
+		if hud_inc in(1 2 3) then max_ocost=HHINCOME_a/12*.3; *under 80% of AMI then pay 30% threshold;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: see above comment in renter section.

File: Prog/NCHousing_needs_units_targets.sas [hunk 10]
```diff
@@ -0,0 +1,927 @@
...
+		else do; 
+		    if BUILTYR2  in (07, 08, 09, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22) then structureyear=1;
+			else if BUILTYR2  in (04, 05, 06) then structureyear=2;
+            else if BUILTYR2 in (01, 02, 03)  then structureyear=3;
+		end;
+
+		
+	if rentgrs in ( 9999999, .n , . ) then affordable=.;
+		else do; 
+		    if rentgrs_a<750 then affordable=1;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: this may need to be adjusted.

File: Prog/NCHousing_needs_units_targets.sas [hunk 11]
```diff
@@ -0,0 +1,927 @@
...
+data projectedHH;
+set NChsg.projection_for_calibration;
+county2_char= puma;
+run;
+
+data fiveyeartotal1;
+set Housing_needs_baseline_2013_3 Housing_needs_baseline_2014_3 Housing_needs_baseline_2015_3 Housing_needs_baseline_2016_3 Housing_needs_baseline_2017_3;
+totalpop=0.2;
+merge=1;
+totpop_wt= totalpop*AFACT2;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: we need to remove afact 2 here right?

File: Prog/NCHousing_needs_units_targets.sas [hunk 12]
```diff
@@ -0,0 +1,927 @@
...
+    sum='Sum of HHWTs' * ( hhwt_geo='Original 5-year'  )
+  / box='Occupied housing units';
+  *format county2_char county2_char.;
+run;
+
+data fiveyeartotal_vacant;
+	set Housing_needs_vacant_2013 Housing_needs_vacant_2014 Housing_needs_vacant_2015 Housing_needs_vacant_2016 Housing_needs_vacant_2017;
+totalpop=0.2;
+merge=1;
+totpop_wt= totalpop*AFACT2;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: delete afact2?

File: Prog/NCHousing_needs_units_targets.sas [hunk 13]
```diff
@@ -0,0 +1,927 @@
...
+  / box='Vacant (nonseasonal) housing units';
+  *format county2_char county2_char.;
+run;
+
+/*need to account for other vacant units in baseline and future targets for the region to complete picture of the total housing stock*/
+data fiveyeartotal_othervacant;
+   set other_vacant_2013 other_vacant_2014 other_vacant_2015 other_vacant_2016 other_vacant_2017;
+totalpop=0.2;
+merge=1;
+totpop_wt= totalpop*AFACT2;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: delete afact2?

File: Prog/NCHousing_needs_units_targets.sas [hunk 14]
```diff
@@ -0,0 +1,927 @@
...
+hhwt_geo=.; 
+
+hhwt_geo=hhwt*calibration*0.2; 
+
+label hhwt_geo="Household Weight Calibrated to Steven Estimates for Households"
+	  calibration="Ratio of Steven 2015 estimate to ACS 2013-17 for 45 geographic units";
+
+run; 
+
+proc univariate data= fiveyeartotal_c;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: does this result match the one from the household income projections program?

File: Prog/NCHousing_needs_units_targets.sas [hunk 15]
```diff
@@ -0,0 +1,927 @@
...
+%single_year(2015); 
+%single_year(2016);
+%single_year(2017);
+
+/*merge single year data and reweight
+
+revised to match Steven's files in https://urbanorg.box.com/s/nm4wb0arxsxv8aml98b6y062tudzp2i6
+*/
+data projectedHH;
+set NChsg.projection_for_calibration;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: did you need to rerun this based on the adjustments that were made today?

File: Prog/NCHousing_needs_units_targets.sas [hunk 16]
```diff
@@ -0,0 +1,927 @@
...
+		if P_60  =< hhincome_a < P_80 then inc=4;
+		if P_80  =< hhincome_a < P_100 then inc=5;
+  end;
+	    label /*hud_inc = 'HUD Income Limits category for household (2016)'*/
+	    inc='Income quintiles statewide not account for HH size';
+		format inc inc_cat.; 
+run;
+
+/*export dataset*/
+ data NCHsg.fiveyeartotal;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: add data set label please and run a proc contents.

File: Prog/NCHousing_needs_units_targets.sas [hunk 17]
```diff
@@ -0,0 +1,927 @@
...
+run;
+
+/*export dataset*/
+ data NCHsg.fiveyeartotal; 
+   set fiveyeartotal;
+ run;
+
+proc tabulate data=fiveyeartotal format=comma12. noseps missing;
+  class county2_char;
+  var hhwt_geo;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: add original 5 year weights - hhwt*.2.

File: Prog/NCHousing_needs_units_targets.sas [hunk 18]
```diff
@@ -0,0 +1,927 @@
...
+ data NCHsg.fiveyeartotal_vacant; 
+   set fiveyeartotal_vacant_c;
+ run;
+
+proc tabulate data=fiveyeartotal_vacant_c format=comma12. noseps missing;
+  class county2_char;
+  var hhwt_geo;
+  table
+    all='Total' county2_char=' ',
+    sum='Sum of HHWTs' * ( hhwt_geo='Original 5-year')
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: should read hhwt_geo='Adjusted to 2015 estimate'

File: Prog/NCHousing_needs_units_targets.sas [hunk 19]
```diff
@@ -0,0 +1,927 @@
...
+
+run; 
+
+/*export dataset*/
+ data NCHsg.fiveyeartotal_vacant; 
+   set fiveyeartotal_vacant_c;
+ run;
+
+proc tabulate data=fiveyeartotal_vacant_c format=comma12. noseps missing;
+  class county2_char;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: add original 5 year  - as hhwt*.2

File: Prog/NCHousing_needs_units_targets.sas [hunk 20]
```diff
@@ -0,0 +1,927 @@
...
+ data NCHsg.fiveyeartotal; 
+   set fiveyeartotal;
+ run;
+
+proc tabulate data=fiveyeartotal format=comma12. noseps missing;
+  class county2_char;
+  var hhwt_geo;
+  table
+    all='Total' county2_char=' ',
+    sum='Sum of HHWTs' * ( hhwt_geo='Original 5-year'  )
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: should read hhwt_geo='Adjusted to 2015 estimate'

File: Prog/NCHousing_needs_units_targets.sas [hunk 21]
```diff
@@ -0,0 +1,927 @@
...
+
+hhwt_geo=hhwt*calibration*0.2; 
+
+label hhwt_geo="Household Weight Calibrated to Steven Estimates for Households"
+	  calibration="Ratio of Steven 2015 estimate to ACS 2013-17 for 45 geographic units";
+
+run; 
+
+/*export dataset*/
+ data NCHsg.fiveyeartotal_othervacant;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: add data set label and run proc contents

File: Prog/NCHousing_needs_units_targets.sas [hunk 22]
```diff
@@ -0,0 +1,927 @@
...
+run; 
+
+/*export dataset*/
+ data NCHsg.fiveyeartotal_othervacant; 
+   set fiveyeartotal_othervacant_c;
+ run;
+
+proc tabulate data=fiveyeartotal_othervacant_c format=comma12. noseps missing;
+  class county2_char;
+  var hhwt_geo;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: see above comments on this proc tabulate.

File: Prog/NCHousing_needs_units_targets.sas [hunk 23]
```diff
@@ -0,0 +1,927 @@
...
+
+hhwt_geo=hhwt*calibration*0.2; 
+
+label hhwt_geo="Household Weight Calibrated to Steven Estimates for Households"
+	  calibration="Ratio of Steven 2015 estimate to ACS 2013-17 for 45 geographic units";
+
+run; 
+
+/*export dataset*/
+ data NCHsg.fiveyeartotal_vacant;
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets.sas]: add data set label and run proc contents.

File: Prog/NCHousing_needs_units_targets - Alt.sas [hunk 24]
```diff
@@ -0,0 +1,742 @@
+/**************************************************************************
+ Program:  Housing_needs_units_targets-Alt.sas
```
Comment (lhendey) [inline Prog/NCHousing_needs_units_targets - Alt.sas]: please replicate any comments from the main program.

File: Prog/NC target tables tabulation.sas [hunk 25]
```diff
@@ -0,0 +1,537 @@
...
+
+ data fiveyeartotal_vacant; 
+   set NCHsg.fiveyeartotal_vacant;
+ run;
+
+ data fiveyeartotal_othervacant; 
+   set NCHsg.fiveyeartotal_othervacant ;
+ run;
+
+proc tabulate data=fiveyeartotal format=comma12. noseps missing;
```
Comment (lhendey) [inline Prog/NC target tables tabulation.sas]: remove these, they are in the other program.

File: Prog/NC target tables tabulation.sas [hunk 26]
```diff
@@ -0,0 +1,537 @@
...
+  class county2_char;
+  var hhwt_5;
+  table
+    all='Total' county2_char=' ',
+    sum='Sum of HHWTs' * ( hhwt_5='Original 5-year' )
+  / box='Seasonal vacant housing units';
+  *format county2_char county2_char.;
+run;
+
+proc freq data=fiveyeartotal_othervacant;
```
Comment (lhendey) [inline Prog/NC target tables tabulation.sas]: this is fine here but i think it should be removed from the other program.

File: Prog/NC target tables tabulation.sas [hunk 27]
```diff
@@ -0,0 +1,537 @@
...
+*format county2_char county2_char.;
+run; 
+proc export data=fiveyeartotal_othervacant
+ 	outfile="&_dcdata_default_path\NCHsg\Prog\other_vacant_&date..csv"
+   dbms=csv
+   replace;
+   run;
+
+/*data set for all units that we can determine cost level*/ 
+data all;
```
Comment (lhendey) [inline Prog/NC target tables tabulation.sas]: please create "all" in the other program and add a data set label and proc contents.

File: Prog/NC target tables tabulation_alt.sas [hunk 28]
```diff
@@ -0,0 +1,630 @@
+/**************************************************************************
```
Comment (lhendey) [inline Prog/NC target tables tabulation_alt.sas]: see comments on non alt program.

File: Prog/NC target tables tabulation.sas [hunk 29]
```diff
@@ -0,0 +1,537 @@
...
+	set fiveyeartotal fiveyeartotal_vacant (in=a);
+	if a then inc=6; 
+format inc inc_cat.;
+run; 
+
+/*output current households by unit cost catgories by tenure*/
+proc freq data=all;
+tables inc*allcostlevel /nopercent norow nocol out=region_units;
+weight hhwt_5;
```
Comment (lhendey) [inline Prog/NC target tables tabulation.sas]: this code hasn't been updated to reflect your new vars for inc and hhwt_5.

File: Prog/NC target tables tabulation.sas [hunk 30]
```diff
@@ -0,0 +1,537 @@
+/**************************************************************************
+ Program:  NC target tables tabulation.sas
```
Comment (lhendey) [inline Prog/NC target tables tabulation.sas]: this isn't running clean on a batch submit.

File: Prog/NCHousing_needs_units_targets.sas (hunk 7 above)
Comment (maggiesu0725) [reply Prog/NCHousing_needs_units_targets.sas]: how did you came up with the $443 number?

Comment (lhendey) [reply Prog/NCHousing_needs_units_targets.sas]: hmm not sure. use the family of 4 categories to come up with some ranges. don't have to match precisely.

File: Prog/NCHousing_needs_units_targets.sas (hunk 15 above)
Comment (maggiesu0725) [reply Prog/NCHousing_needs_units_targets.sas]: yes it is updated.

File: Prog/NCHousing_needs_units_targets.sas (hunk 9 above)
Comment (maggiesu0725) [reply Prog/NCHousing_needs_units_targets.sas]: The new costlevel cutoffs are 350, 700, 1050, 1500, 2500.

File: Prog/NCHousing_needs_units_targets.sas (hunk 10 above)
Comment (maggiesu0725) [reply Prog/NCHousing_needs_units_targets.sas]: According to the new costlevel cutoffs are 350, 700, 1050, 1500, 2500, this should be 350? @lhendey

File: Prog/NCHousing_needs_units_targets.sas (hunk 14 above)
Comment (maggiesu0725) [reply Prog/NCHousing_needs_units_targets.sas]: that one is not calibrated, and this is calibrated using uncalibrated tabulation and projection outcome. @lhendey

**Comment by maggiesu0725** [comment]
could you review just NCHousing_needs_units_targets.sas and if the new categories look oK I will update the alt program and the tabulation programs? @lhendey

**Comment by lhendey** [comment]
@maggiesu0725 - i think you need to push your recent commits.

File: Prog/NCHousing_needs_units_targets.sas (hunk 14 above)
Comment (lhendey) [reply Prog/NCHousing_needs_units_targets.sas]: @maggiesu0725 i'm not following your response.

File: Prog/NCHousing_needs_units_targets.sas (hunk 10 above)
Comment (lhendey) [reply Prog/NCHousing_needs_units_targets.sas]: @maggiesu0725 no - we didn't use the bottom category the last time - we used the second - (which was $1300) so use $700?

File: Prog/NCHousing_needs_units_targets.sas (hunk 9 above)
Comment (lhendey) [reply Prog/NCHousing_needs_units_targets.sas]: @maggiesu0725 okay - you need to update the formats up top.

File: Prog/NC target tables tabulation.sas (hunk 30 above)
Comment (maggiesu0725) [reply Prog/NC target tables tabulation.sas]: will move this to issue 7.

**Pull request merged**